streams/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java

```diff
@@ -90,24 +91,64 @@ public void setup() {
         );
     }
 
+    private void startStreamsAndWaitForRunning(final KafkaStreams kafkaStreams) throws InterruptedException {
+        kafkaStreams.start();
+        waitForStateTransition(KafkaStreams.State.RUNNING);
+    }
+
     @After
     public void teardown() throws IOException {
+        stateTransitionHistory.clear();
         purgeLocalStreamsState(properties);
     }
 
+    private void addStreamStateChangeListener(final KafkaStreams kafkaStreams) {
+        // we store each new state in state transition so that we won't miss any state change
+        kafkaStreams.setStateListener(
+            (newState, oldState) -> stateTransitionHistory.add(newState)
+        );
+    }
+
+    private void waitForStateTransition(final KafkaStreams.State expected) throws InterruptedException {
+        waitForCondition(
+            () -> !stateTransitionHistory.isEmpty() && stateTransitionHistory.contains(expected),
+            DEFAULT_DURATION.toMillis(),
+            () -> String.format("Client did not change to the %s state in time. Observed new state transitions: %s",
+                expected, stateTransitionHistory)
+        );
+    }
+
+    // verify if there's the state change from "before" state into "after" state
+    private boolean hasStateTransition(final KafkaStreams.State before, final KafkaStreams.State after) {
+        // should have at least 2 states in history
+        if (stateTransitionHistory.size() < 2) {
+            return false;
+        }
+
+        for (int i = 0; i < stateTransitionHistory.size() - 1; i++) {
+            if (stateTransitionHistory.get(i).equals(before) && stateTransitionHistory.get(i + 1).equals(after)) {
+                return true;
+            }
+        }
+        return false;
```

Comment:
Why do we need a `for`-loop here? Wouldn't it suffice to verify the last two elements of the history and check if those two elements are a `REBALANCING` followed by a `RUNNING`?

Reply:
I use `for loop` is because I think there could be cases that there are some other state changes after RUNNING, ex: DEAD. But after your question, I think if that happened, the test should also fail as well. So, check the last 2 elements is good. Updated. Thanks.